jni/hidesu.c

```diff
@@ -20,29 +22,266 @@ typedef unsigned short int sa_family_t;
 #include <sys/glibc-syscalls.h>
 #include <asm/unistd.h>
 #include <strings.h>
+#include <dirent.h>
+#include <sys/types.h>
+#include <sys/xattr.h>
+
+char zygote_mnt_ns[256], zygote64_mnt_ns[256], init_mnt_ns[256];
 
 //WARNING: Calling this will change our current namespace
 //We don't care because we don't want to run from here anyway
-int disableSu(int pid) {
-	char *path = NULL;
-	asprintf(&path, "/proc/%d/ns/mnt", pid);
-	int fd = open(path, O_RDONLY);
+int disableSu(unsigned int pid) {
+	char buffer[512];
+	char *p, *p2;
+	struct stat sb;
+	int ret = 0;
+	snprintf(buffer, sizeof(buffer), "/proc/%u/ns/mnt", pid);
+
+	// check that process is running and wait untit it has own mount namespace
+	if(stat(buffer, &sb) == -1) return 1;
+	while(isStillZygoteNs(buffer) != 0) {
+		kill(pid, SIGCONT);
+		usleep(20);
+		kill(pid, SIGSTOP);
+	}
```

Comment:
If root would get detected, then the too long delay could also have happened before first SIGSTOP (during logcat reading etc.). I think detecting the new process could be our bottleneck (if root would get detected)...